File: Hard Challenges/Challenge #0001 [Hard]/solutions/solution.rb
class NumberGuesser
  MIN_ANSWER = 1
  MAX_ANSWER = 100

  def initialize
    @current_min = MIN_ANSWER
    @current_max = MAX_ANSWER
    @guesses = 1
  end

  def guess_until_found
    loop do
      perform_sanity_check
      case make_guess
      when '1'
        @current_min = current_guess + 1
        @guesses += 1
      when '2'
        @current_max = current_guess - 1
        @guesses += 1
      when '3'
        puts "Woohoo! I guessed your number in #{@guesses} tries!"
        break
      else
        puts 'Please enter either "1", "2" or "3".'
      end
    end
  end

  private

  def make_guess
    puts "My guess is: #{current_guess}"
    puts 'Was I right? 1) higher, 2) lower, 3) YES!'
    gets.chomp
  end

  def current_guess
    (@current_max + @current_min)/2
  end

  def perform_sanity_check
    if @current_max < @current_min
      puts 'Oops, something impossible has happened -- Looks like you cheated!'
      exit
    end
  end
end

puts '***** Number guessing game!! *****'
puts "Pick a number between #{NumberGuesser::MIN_ANSWER}-#{NumberGuesser::MAX_ANSWER}"
puts
sleep 2

NumberGuesser.new.guess_until_found

File: Hard Challenges/Challenge #0001 [Hard]/solutions/solution.py
class Number_Guesser(object):
	lower_bound = 1
	upper_bound = 100

	def __init__(self,lower_bound,upper_bound):
		self.lower_bound = lower_bound
		self.upper_bound = upper_bound

	def guess_and_adjust(self):
		number_of_guesses = 1
		while True:
			self.anticheat()
			my_guess = self.make_guess()
			answer = raw_input('Enter 1 if your number is higher, 2 if it\'s lower, or 3 if I was correct!\n>')
			if answer=='1':
				self.lower_bound = my_guess + 1
				number_of_guesses = number_of_guesses + 1
			elif answer=='2':
				self.upper_bound = my_guess - 1
				number_of_guesses = number_of_guesses + 1
			elif answer=='3':
				print 'Woohoooo! I guessed your number in {0} guesses!'.format(number_of_guesses)
				quit()

	def anticheat(self):
		if self.lower_bound > self.upper_bound:
			print 'You cheated! We\'re done here!'
			quit()

	def make_guess(self):
		self.current_guess = (self.upper_bound + self.lower_bound)/2
		print 'my guess is {0}'.format(self.current_guess)
		return self.current_guess

n = Number_Guesser(1,100)
n.guess_and_adjust()
